### ISSUE TITLE
[no-useless-constructor] False positive on protected constructors

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#4772]
_I skipped parts of this template because the issue is reproducible on the playground._

`no-useless-constructor` is complaining about a "useless constructor" when shadowing a `protected` constructor. Following the code means breaking typescript, which then says

```
Constructor of class 'A' is protected and only accessible within the class declaration.(2674)
```

## Repro

https://typescript-eslint.io/play#ts=4.6.2&sourceType=module&code=MYGwhgzhAECC0G8BQ1oAcBOB7ALgU2HwBNpgsA7CHDAV0KwwAoBbPKMAczwC5oqMAluQ4BKRClSkKELCDwA6EFg4s2ETnhEBuCQF8k+pKEgwAQtDwAPfOSIx4yVGUrU6OBqvZde-IaPGSfDRoeEysXpo6qPqG5HgA7tCmjADk4epcKSJIQA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYCuCcuiCWSZBCEKlSEZKIAXCHBQpuIAL5A&tsConfig=N4XyA

```JSON
{
  "rules": {
    "@typescript-eslint/no-useless-constructor": "error"
  }
}
```

```TS
class A {
  protected constructor(message: string) {
    console.log(message);
  }
}

class B extends A {
  constructor(message: string) {
    super(message);
  }
  // ^-- Useless constructor.ESLint(@typescript-eslint/no-useless-constructor)
}

new B('message')
```

## After the "fix"

❌ TypeScript will not complete successfully:

```TS
class A {
  protected constructor(message: string) {
    console.log(message);
  }
}

class B extends A {}

new B('message')
// ^-- Constructor of class 'A' is protected and only accessible within the class declaration.(2674)
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.17.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.1.0` |

**Comment by bradzacher** [comment]
This is working as expected. The rule does not use type information - thus it cannot determine that the parent class has a different accessibility.

This extension rule just adds support for some TS-specific syntax.

We will just need to accept this false-positive as a rare edge-case. Given this rule doesn't have a fixer - this is acceptable.

**Comment by fregante** [comment]
> The rule does not use type information

Isn't this the repo of `typescript-eslint`? Why does the rule not use type information? What is the point of overriding the native `no-useless-constructor` with `@typescript-eslint/no-useless-constructor` if it doesn't use type information? Sorry this makes no sense to me.

**Comment by fregante** [comment]
From the readme:

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-useless-constructor.md#rule-details

> This rule extends the base [eslint/no-useless-constructor](https://eslint.org/docs/rules/no-useless-constructor) rule. It adds support for:
> 
> - constructors marked as protected / private (i.e. marking a constructor as non-public),

This rule is literally _not_ working as expected 🤷‍♂️

**Comment by bradzacher** [comment]
You misunderstand.
That support is for `private constructor() {}`

---

Actually reviewing it, you can mark the constructor as necessary by adding an explicit accessibility.

Explicit public along with a super class is allowed by the rule.

Using implicit public accessibility is ambiguous and requires type information. Which is not possible to support.

**Comment by fregante** [comment]
> requires type information

Isn't this typescript-eslint? Is this information available?

**Comment by bradzacher** [comment]
> Is there any way to exclude this constructor

You can add an explicit `public` to the constructor.

----

>  Is this information available?

In some projects, yes. In most - no.
Adding type info is a major breaking change because it precludes much of our user-base from being able to use a rule.

Esp in the case of an extension rule the question is - is it worth it?
Modifying base class constructor visibility with an implicit public constructor is a very rare usecase that has a workaround built into the rule.

Thus in our opinion - it is not worth the massive cost to add type info to fully handle this rare case.

**Comment by fregante** [comment]
I found another exception:

```js
export class MySpecificError extends Error {
  override name = "MySpecificError";
  override readonly cause: DOMException;
  // eslint-disable-next-line @typescript-eslint/no-useless-constructor -- Required to make the types stricter
  constructor(message: string, options: { cause: DOMException }) {
    super(message, options);
  }
}
```

Once again I'm using the constructor types to enforce a specific type, without changing the implementation. The underlying JavaScript constructor is indeed useless, but in TypeScript it's useful.

https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=KYDwDg9gTgLgBAYwDYEMDOa4FkCeBlMYBASwDNiEBRKKaOUGYAOwBNNrao4BvAKDjgQAbsBrEWwOExQBbSQF44AIlwEiZCh2hKA3P0EixEuFGAoWEJkhyIUAVzTAAXHAAiAeSyUQCYGBjElnoCAPQh9GhIxEwwALQsxGgoAEZIwLFMDLFRmXAAAjA4hGgIUMT+scCR0TAhTBCxDsBpGLEIlmgwUHYIMHSxsXAASsAAjnbEpixwfXAyKADWkjAAFstFVXCdZb2i+u1M2z19UAAUchgoAObOW13RVwA0gv6Bhy7ctk0uHl4+fgFLHAAL4ASh4+gEaDshDOFySN2eEFeHVBwRBvGBvF4mQA7th8IQSOQqDRoKcAOTw67AClouC8MJwbxExjTABMcBQUCudjkMTQz2SdngVwg8AAjAA6U7sgCscoALKDsXiCWpiZoyWcqVUEbTntwEPZHC41VodQB1WhMK4zDZ0sF6JkAFQ2cApFopcESc0SaAeMzWcFIECQSAguMDYFosICm1ItBk9sIHt+3l8KKYFNVwHxqiJGlJnEp1JuFMNxu+Ujzbk8GYBb0pAGFoKZeinaaCnYzwoBQciAA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYCuCcuiCWSZBCEKlSEZKIAXCHBQpuIAL5A&tsConfig=N4XyA
